pkg/health/health.go

```diff
@@ -0,0 +1,133 @@
+// Package health provides the mechanism to report the health status of the service
+package health
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	boxLength       = time.Minute
+	totalBoxes      = 5
+	windowLength    = totalBoxes * time.Minute
+	totalMetricsTag = "Total"
+)
+
+// mockable clock provider
+var clock = time.Now
+
+// Status represents the different states that a service can have
+type Status int
+
+const (
+	// Starting the service. It is not yet ready to receive requests
+	Starting Status = iota
+	// Ready service. It is healthy and ready to receive requests
+	Ready
+	// Error status. The service is not healthy
+	Error
+)
+
+func (s Status) String() string {
+	switch s {
+	case Starting:
+		return "STARTING"
+	case Ready:
+		return "READY"
+	case Error:
+		return "ERROR"
+	default:
+		return "UNKNOWN"
+	}
+}
+
+// Reporter of the health status and the statistics of the service
+type Reporter struct {
+	Status        Status
+	recentMetrics metricsWindow
+	totalMetrics  MetricSet
+}
+
+// Metrics return the current statistics of the service. It returns 1m, 5m and total (all time) metrics
+func (r *Reporter) Metrics() map[string]MetricSet {
+	aggr := MetricSet{}
+
+	r.recentMetrics.mut.Lock()
+	boxes := r.recentMetrics.boxes
+	r.recentMetrics.mut.Unlock()
+	now := clock()
+
+	var recent MetricSet
+	if len(boxes) > 0 {
+		recent = boxes[len(boxes)-1].metrics
+	}
+	for _, cb := range boxes {
+		if now.Sub(cb.start) >= windowLength {
+			continue
+		}
+		aggr.ReceivedFlows += atomic.LoadUint64(&cb.metrics.ReceivedFlows)
+		aggr.DiscardedFlows += atomic.LoadUint64(&cb.metrics.DiscardedFlows)
+	}
+
+	return map[string]MetricSet{
+		totalMetricsTag: r.totalMetrics,
+		fmt.Sprintf("%0.fm", boxLength.Minutes()):                recent,
+		fmt.Sprintf("%0.fm", (totalBoxes * boxLength).Minutes()): aggr,
+	}
+}
+
+// FlowProcessed accounts the event of a flow being successfully processed
+func (r *Reporter) FlowProcessed() {
+	cb := r.recentMetrics.currentBox()
+	atomic.AddUint64(&cb.metrics.ReceivedFlows, 1)
+	atomic.AddUint64(&r.totalMetrics.ReceivedFlows, 1)
+}
+
+// FlowDiscarded accounts the event of a flow being discarded due to an error
+// (e.g. the flow couldn't be decoded)
+func (r *Reporter) FlowDiscarded() {
+	cb := r.recentMetrics.currentBox()
+	atomic.AddUint64(&cb.metrics.ReceivedFlows, 1)
+	atomic.AddUint64(&cb.metrics.DiscardedFlows, 1)
+	atomic.AddUint64(&r.totalMetrics.ReceivedFlows, 1)
+	atomic.AddUint64(&r.totalMetrics.DiscardedFlows, 1)
+}
+
+type metricsWindow struct {
+	mut   sync.RWMutex
+	boxes []*metricsBox
+}
+
+func (sw *metricsWindow) currentBox() *metricsBox {
+	sw.mut.Lock()
+	defer sw.mut.Unlock()
+	if len(sw.boxes) == 0 ||
+		clock().Sub(sw.boxes[len(sw.boxes)-1].start) >= boxLength {
+		sw.pushBox()
+	}
+	return sw.boxes[len(sw.boxes)-1]
```

Comment:
That's smart ! 

But does it mean if there is no flows coming for a time (typically hanging at `r.format.Next()`) you will not push boxes ?

Reply:
You are right, but it should not matter since in the stats aggregation we just count the boxes whose start time is less than 5 minutes.